include/mxnet/ndarray.h

```diff
@@ -90,10 +90,20 @@ class NDArray {
    * \param ctx context of NDArray
    * \param delay_alloc whether delay the allocation
    * \param dtype data type of this ndarray
+   * \param name name of this ndarray
    */
   NDArray(const mxnet::TShape &shape, Context ctx,
-          bool delay_alloc = false, int dtype = mshadow::default_type_flag)
-      : ptr_(std::make_shared<Chunk>(shape, ctx, delay_alloc, dtype)),
+          bool delay_alloc = false,
+          int dtype = mshadow::default_type_flag
+#if MXNET_ENABLE_STORAGE_TAGGING
+        , const std::string& name = MXNET_DEFAULT_STORAGE_TAG("unknown")
+#endif  // MXNET_ENABLE_STORAGE_TAGGING
```

Comment:
can we just default as nullptr and use #if inside the function definition. same at other places.

Reply:
The reason why the default name is set to the `MXNET_DEFAULT_STORAGE_TAG` instead of `""` is because `MXNET_DEFAULT_STORAGE_TAG` defaults the string value to the **function name of the caller**.

This is important for avoiding making intrusive changes inside the operator implementations. For instance, as we have shown above, each MXNet operator can potentially request temporary resource, without the `MXNET_DEFAULT_STORAGE_TAG` macro that allows us to trace back to the caller, we need to annotate the resource allocations inside each operator, which is obviously not a very scaleable solution.